lib/Conversion/DCToHW/DCToHW.cpp

```diff
@@ -0,0 +1,946 @@
+
+#include "circt/Conversion/DCToHW.h"
+#include "../PassDetail.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/DC/DCDialect.h"
+#include "circt/Dialect/DC/DCOps.h"
+#include "circt/Dialect/DC/DCPasses.h"
+#include "circt/Dialect/ESI/ESIOps.h"
+#include "circt/Dialect/ESI/ESITypes.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWTypes.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "circt/Support/BackedgeBuilder.h"
+#include "circt/Support/ConversionPatterns.h"
+#include "circt/Support/ValueMapper.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+#include "llvm/Support/MathExtras.h"
+
+#include <optional>
+
+#include "llvm/Support/Debug.h"
+
+using namespace mlir;
+using namespace circt;
+using namespace circt::dc;
+using namespace circt::hw;
+
+using NameUniquer = std::function<std::string(Operation *)>;
+
+// NOLINTNEXTLINE(misc-no-recursion)
+static Type tupleToStruct(TupleType tuple) {
+  auto *ctx = tuple.getContext();
+  mlir::SmallVector<hw::StructType::FieldInfo, 8> hwfields;
+  for (auto [i, innerType] : llvm::enumerate(tuple)) {
+    Type convertedInnerType = innerType;
+    if (auto tupleInnerType = innerType.dyn_cast<TupleType>())
+      convertedInnerType = tupleToStruct(tupleInnerType);
+    hwfields.push_back({StringAttr::get(ctx, "field" + std::to_string(i)),
+                        convertedInnerType});
+  }
+
+  return hw::StructType::get(ctx, hwfields);
+}
+
+/// Converts the range of 'types' into a `hw`-dialect type. The range will be
+/// converted to a `hw.struct` type.
+// NOLINTNEXTLINE(misc-no-recursion)
+static Type toHWType(Type t);
+static Type toHWType(TypeRange types) {
+  if (types.size() == 1)
+    return toHWType(types.front());
+  return toHWType(mlir::TupleType::get(types[0].getContext(), types));
+}
+
+/// Converts any type 't' into a `hw`-compatible type.
+/// tuple -> hw.struct
+/// none -> i0
+/// (tuple[...] | hw.struct)[...] -> (tuple | hw.struct)[toHwType(...)]
+// NOLINTNEXTLINE(misc-no-recursion)
+static Type toHWType(Type t) {
+  return TypeSwitch<Type, Type>(t)
+      .Case([&](TupleType tt) { return toHWType(tupleToStruct(tt)); })
+      .Case([&](hw::StructType st) {
+        llvm::SmallVector<hw::StructType::FieldInfo> structFields(
+            st.getElements());
+        for (auto &field : structFields)
+          field.type = toHWType(field.type);
+        return hw::StructType::get(st.getContext(), structFields);
+      })
+      .Case([&](NoneType nt) { return IntegerType::get(nt.getContext(), 0); })
+      .Default([&](Type t) { return t; });
+}
+
+static Type toESIHWType(Type t) {
+  auto *ctx = t.getContext();
+  Type outType =
+      llvm::TypeSwitch<Type, Type>(t)
+          .Case<ValueType>([&](auto vt) {
```

Comment:
NIT: As in the previous `TypeSwitch`, the template parameters can be inferred from the lambda parameter type, if it's specified.